File: KalmanFilter.py
class KalmanFilter(object):
	q = 0.0
	r = 0.0
	x = 0.0
	p = 0.0
	k = 0.0

	def __init__(self, x_initial_value, q = 0.1, r = 0.1, p = 0.1, k = 0.0):
		self.q = q
		self.r = r
		self.x = x_initial_value
		self.p = p
		self.k = k

	def update(self, input_i):
		self.p = self.p + self.q
		self.k = self.p / (self.p + self.r)
		self.p = (1 - self.k) * self.p
		self.x = self.x + self.k * (input_i - self.x)
		return self.x

File: Lab3/Sources/Kalmanfilter.c
#include <stdio.h>
#include "Kalmanfilter.h"


int Kalmanfilter_C(float* InputArray, float* OutputArray, kalman_state* kstate, int Length) {		
	float x_predict, p_predict;
	int i;
	
	for(i = 0; i < Length; i++) {		
		// Eq'1
		x_predict = kstate->f * kstate->x;
		// kstate->x = kstate->f * xOld;
		
		// Eq'2
		p_predict = kstate->f * kstate->f * kstate->p + kstate->q;
		//kstate->p = kstate->f * kstate->f * x_predict + kstate->q;
		
		// Eq'3
		kstate->k = (p_predict * kstate->h) / (kstate->h * kstate->h * p_predict + kstate->r);
		
		// Eq'4
		kstate->p = (1 - kstate->k * kstate->h) * p_predict;
		
		// Update x
		kstate->x = x_predict + kstate->k * (InputArray[i] - kstate->h * x_predict);
		
		// Eq'5
		OutputArray[i] = kstate->x;
		
	}
	
	return 0;
}
